### PR TITLE
the general rpmsg drivers

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
rpmsg-syslog redirect log from remote processor
rpmsg-tty redirect console from remote processor
rpmsg-clk redirect clk API from remote processor
rpmsg-hostfs redirect fs API from remote processor
rpmsg-rtc redirect rtc API from remote processor
rpmsg-usrsock redirect socket API from remote processor
With these drivers, the remote processor could access many common resource on the host side.

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
@lgirdwood

**Comment by lgirdwood** [comment]
@xiaoxiang781216 sorry, it's best to always ping reviewer or add me if no reply in 1 day otherwise no one is notified of a new PR. I've also added Keyon for review too. Will review on Monday.

File: drivers/rpmsg/rpmsg_syslog.c [hunk 1]
```diff
@@ -0,0 +1,157 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (C) 2017 Pinecone Inc.
```
Comment (lgirdwood) [inline drivers/rpmsg/rpmsg_syslog.c:3]: Some description here for what services this file provides.

File: drivers/rpmsg/rpmsg_syslog.c [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+#define rpmsg_syslog_resume		rpmsg_syslog_header
+#define rpmsg_syslog_transfer_done	rpmsg_syslog_header
+
+struct rpmsg_syslog {
+	char				*tmpbuf;
+	unsigned int			nextpos;
+	unsigned int			alloced;
+};
+
+static int rpmsg_syslog_callback(struct rpmsg_device *rpdev,
```
Comment (lgirdwood) [inline drivers/rpmsg/rpmsg_syslog.c:38]: some comments needed here explaining what this does

File: drivers/rpmsg/rpmsg_syslog.c [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		unsigned int printed = 0;
+		const char *nl;
+
+		nl = memrchr(msg->data, '\n', msg->count);
+		if (nl != NULL) {
+			printed = nl + 1 - msg->data;
+			copied = msg->count - printed;
+
+			if (priv->nextpos) {
+				pr_info(KERN_TIME "%.*s%.*s", priv->nextpos,
```
Comment (lgirdwood) [inline drivers/rpmsg/rpmsg_syslog.c]: dev_info or dev_dbg ?

File: drivers/clk/clk.c [hunk 4]
```diff
@@ -344,6 +344,7 @@ unsigned int __clk_get_enable_count(struct clk *clk)
 {
 	return !clk ? 0 : clk->core->enable_count;
 }
+EXPORT_SYMBOL_GPL(__clk_get_enable_count);
```
Comment (lgirdwood) [inline drivers/clk/clk.c]: Clock enable count should not be known by clients, if you need this it sometimes implies you are using the API incorrectly.

File: fs/file.c [hunk 5]
```diff
@@ -957,6 +957,7 @@ int f_dupfd(unsigned int from, struct file *file, unsigned flags)
 	}
 	return err;
 }
+EXPORT_SYMBOL(f_dupfd);
```
Comment (lgirdwood) [inline fs/file.c]: These three will probably require strong justification for each.

**Comment by keyonjie** [comment]
@xiaoxiang781216 I will spend some time to study and try to give comments, before that, can you please fix those checkpatch issues:
https://download.01.org/0day-ci/sof/kernel-checkpatch-pr-177-1489e1.txt

**Comment by lgirdwood** [comment]
@xiaoxiang781216 yep, please fix checkpatch before emailing to LKML

File: drivers/rpmsg/rpmsg_syslog.c (hunk 1 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/rpmsg_syslog.c:3]: Ok, I will add in the next pull.

File: drivers/rpmsg/rpmsg_syslog.c (hunk 3 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/rpmsg_syslog.c]: Since the remote side may already add some prefix(e.g. dsp: or sensor: ), call pr_info instead of dev_info could avoid the prefix flooding.

File: drivers/rpmsg/rpmsg_syslog.c (hunk 1 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/rpmsg_syslog.c:3]: Sure, will add in the new review

**Comment by xiaoxiang781216** [comment]
Ok, I will address checkpatch warnning, in the next review.

File: fs/file.c (hunk 5 above)
Comment (xiaoxiang781216) [reply fs/file.c]: rpmsg-hostfs need call these functions to fully redirect all remote side file related functions, but if we don't make rpmsg-hostfs as module, these function don't expose. How about we change tristate to bool in Kconfig?

File: drivers/rpmsg/rpmsg_syslog.c (hunk 2 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/rpmsg_syslog.c:38]: the code is to handle this case:
the first mesage is "[    0.000000] sensor: NuttX version 7.26-686de48"
the second messgae is "-dirty (arm@song/banks) Oct 12\n"
we have to merge these two message into one to avoid the log interlace with other kernel log.
If the remote side implement the log buffer as a circle buffer, it most likely has to split the buffer and send in two ipc message at the wrapper pointer.

File: fs/file.c (hunk 5 above)
Comment (lgirdwood) [reply fs/file.c]: I'm not sure, this would be upto the clock maintainer and not me. Best to send the patches to LKML after you have made updates from this review.

**Comment by xiaoxiang781216** [comment]
@lgirdwood and @keyonjie, the new patchset address all issue, please review them, thanks.

**Comment by xiaoxiang781216** [comment]
Hi @plbossart,
As discuss with @lgirdwood, the current SOF design couple all IPC(e.g. audio, trace...) in one place which make it's difficult to add more IPC service as need.
For example, sensor hub with voice trigger may need support SOF and sensor, both need IPC to the different kernel subsystem(asoc and iio).
We propose that rpmsg based IPC may the right direction to go, this patchset show that :
1.It's very easy to add new rpmsg service
2.The general service(log, fs, clk, rtc and sock) could be reused by:
  a.The application run on different DSP or MCU, or
  b.The multiple applications run on the same DSP/MCU.
with rpmsg, SOF could just focus on audio part and utilize the general driver do the basic operation(e.g. log, fs, clk, regulator...).
Here is the slide which show how remoteproc/rpmsg could make each subsystem more modularized:
[Audio.pptx](https://github.com/thesofproject/linux/files/2509892/Audio.pptx)

File: include/linux/kern_levels.h [hunk 6]
```diff
@@ -23,6 +23,8 @@
  */
 #define KERN_CONT	KERN_SOH "c"
 
+#define KERN_TIME	KERN_SOH "t"	/* already prefix with the timestamp */
```
Comment (plbossart) [inline include/linux/kern_levels.h]: Neither the commit message nor the definition here are clear. What embedded processor are you talking about and is the timestamp equivalent (could a a different counter or clock source?).

also please fix typo, it's ugly not uglily.

File: lib/string.c [hunk 7]
```diff
@@ -963,6 +963,27 @@ void *memchr(const void *s, int c, size_t n)
...
+ *
+ * returns the address of the last occurrence of @c, or %NULL
+ * if @c is not found
+ */
+void *memrchr(const void *s, int c, size_t n)
+{
+	const unsigned char *p = s + n;
+
+	while (n-- != 0) {
+		if ((unsigned char)c == *--p)
```
Comment (plbossart) [inline lib/string.c:980]: This sort of code is awful. If I need to think to figure out operator precedence and side effects, it's not good in general. Simple is better.

File: drivers/rpmsg/Kconfig [hunk 8]
```diff
@@ -15,6 +15,18 @@ config RPMSG_CHAR
...
 	  receive rpmsg packets.
 
+config RPMSG_SYSLOG
+	tristate "RPMSG syslog redirection"
+	depends on RPMSG
+	help
+	  Say Y here to redirect the syslog/printf from remote processor into
+	  the kernel log which is very useful to see what happen in the remote
+	  side.
+
```
Comment (plbossart) [inline drivers/rpmsg/Kconfig:55]: Not sure I see the benefit from your explanations. if there is already an access to the syslog, pushing it into the kernel log will add more noise. Unless this helps correlate between activity on the remote processor and the main host I don't see the value.

File: drivers/rpmsg/Kconfig [hunk 9]
```diff
@@ -15,6 +15,18 @@ config RPMSG_CHAR
...
+	  If the remote processor hang during bootup or panic at the runtime,
+	  We can even cat /sys/kernel/debug/remoteproc/remoteprocX/trace0 to
+	  get the last log which doesn't output yet.
+
```
Comment (plbossart) [inline drivers/rpmsg/Kconfig:59]: this isn't very clear and tells me that the actual usage is simpler and more complete if you directly access to the trace buffer of the remote processor.

File: drivers/rpmsg/rpmsg_syslog.c [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+ * Copyright (C) 2017 Pinecone Inc.
+ *
+ * redirect syslog/printf from remote to the kernel.
+ */
+
+#include <linux/module.h>
+#include <linux/rpmsg.h>
+#include <linux/slab.h>
+
+#define NOTIMESTAMP			KERN_TIME
```
Comment (plbossart) [inline drivers/rpmsg/rpmsg_syslog.c]: if this definition useful? I personally find it confusing, I see no relationship between notimestamp and kernel_time - it actually seems like a mutually exclusive concept.

File: drivers/rpmsg/rpmsg_syslog.c [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+	dev_set_drvdata(&rpdev->dev, priv);
+	return 0;
+}
+
+static void rpmsg_syslog_remove(struct rpmsg_device *rpdev)
+{
+	struct rpmsg_syslog *priv = dev_get_drvdata(&rpdev->dev);
+
+	if (priv->next)
+		pr_info(NOTIMESTAMP "%.*s\n", priv->next, priv->buf);
```
Comment (plbossart) [inline drivers/rpmsg/rpmsg_syslog.c]: what is this code supposed to do here. Why would one add a print on remove?

File: drivers/rpmsg/rpmsg_core.c [hunk 12]
```diff
@@ -85,6 +85,134 @@ void rpmsg_destroy_ept(struct rpmsg_endpoint *ept)
 }
 EXPORT_SYMBOL(rpmsg_destroy_ept);
 
+/**
```
Comment (plbossart) [inline drivers/rpmsg/rpmsg_core.c:88]: you need to watch out for the software license here. The OpenAmp code is licensed as BSD3 and you are using it in a GPLv2 file. Have you checked with legal that this is fine?

File: drivers/rpmsg/Kconfig [hunk 13]
```diff
@@ -15,6 +15,16 @@ config RPMSG_CHAR
...
+config RPMSG_CLK
+	tristate "RPMSG clock API redirection"
+	depends on COMMON_CLK
+	depends on RPMSG
+	help
+	  Say Y here to redirect clock API from the remote processor.
+	  With this driver, the remote processor could:
+	  1.Reuse the clock driver in the kernel side, or
+	  2.Form a hybrid(kernel plus RTOS) clock tree.
+
```
Comment (plbossart) [inline drivers/rpmsg/Kconfig:27]: I just don't get what this does. No idea...

File: drivers/rpmsg/Kconfig [hunk 14]
```diff
@@ -57,6 +57,15 @@ config RPMSG_SYSLOG
...
 	  get the last log which doesn't output yet.
 
+config RPMSG_USRSOCK
+	tristate "RPMSG socket API redirection"
+	depends on RPMSG
+	depends on NET
+	help
+	  Say Y here to redirect sock API from the remote processor.
+	  This driver could give the remote processor the capability
+	  to access the internet through any host network interface.
```
Comment (plbossart) [inline drivers/rpmsg/Kconfig:67]: I just don't understand this explanation either.

**Comment by xiaoxiang781216** [comment]
Fix the warning addressed by lkp@intel.com

File: drivers/rpmsg/rpmsg_core.c (hunk 12 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/rpmsg_core.c:88]: The concept is from OpenAmp, but the code is written from scratch, is it still a legal issue?

File: drivers/rpmsg/Kconfig (hunk 14 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/Kconfig:67]: Normally, the small DSP/MCU in SoC can't directly access Ethernet/WiFi/Modem hardware and difficult to integrate a full TCP/IP stack.
But, the big core should have some method to access the Internet in general.
With this driver, DSP/MCU could write a simple proxy layer which forward all socket API to the big core and utilize the big core's TCP/IP stack and hardware.
RTOS plus rpmsg-hostfs and rpmsg-usrsock give DSP/MCU a full operating system environment much like Linux.

File: drivers/rpmsg/Kconfig (hunk 13 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/Kconfig:27]: SOF defines a simple clock API here;
sof/src/include/sof/clock.h
Since all clocks normally form a tree structure, the leaf clock frequency decide by:
1.The parent clock frequency
2.The current clock hardware setting
But the parent clock instance normally put into the kernel side, DSP/MCU could use this driver to get the parent clock frequency from kernel to avoid the hard code value.
Sometime we have to change the parent clock frequency, this driver can do so too. 
clk

We can even move MCU/DSP leaf clock definition to kernel side to reuse kernel clock infrastructure as much as possible.

File: drivers/rpmsg/rpmsg_syslog.c (hunk 10 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/rpmsg_syslog.c]: Just to fix checkpatch warning:
WARNING: Possible unnecessary KERN_TIME
#110: FILE: drivers/rpmsg/rpmsg_syslog.c:110:
+               pr_info(KERN_TIME "%.*s\n", priv->next, priv->buf);

Could you accept the code with some reasonable chceckpatch warning?

File: drivers/rpmsg/rpmsg_syslog.c (hunk 11 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/rpmsg_syslog.c]: To avoid the remote log interleave with kernel log, this driver has to buffer the log which isn't terminated by '\n' until the newline received, so the driver need to flush the buffer during unload.

File: drivers/rpmsg/Kconfig (hunk 9 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/Kconfig:59]: Since the buffer in DSP/MCU is limited, rpmsg-syslog client will send the log to kernel regularly to avoid loss the log.
But if DSP/MCU suddenly panic or watchdog reset, some log may still in the buffer, we can get the critical information from /sys/kernel/debug/remoteproc/remoteprocX/trace0 in this case.
syslog

File: drivers/rpmsg/Kconfig (hunk 8 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/Kconfig:55]: syslog/printf here mean the function in the remote processor, not the linux kernel syslog function.:)

File: lib/string.c (hunk 7 above)
Comment (xiaoxiang781216) [reply lib/string.c:980]: The code just copy and modify from memchr in the same file:
void *memchr(const void *s, int c, size_t n)
{
	const unsigned char *p = s;
	while (n-- != 0) {
        	if ((unsigned char)c == *p++) {
			return (void *)(p - 1);
		}
	}
	return NULL;
}

File: include/linux/kern_levels.h (hunk 6 above)
Comment (xiaoxiang781216) [reply include/linux/kern_levels.h]: The embedded processor mean DSP/MCU in SoC,
Yes, the time base on DSP/MCU may different from kernel, but the unit and format is same as kernel.

**Comment by xiaoxiang781216** [comment]
Fix "warning: Variable length array is used." from kbuild test robot, and some comment from @plbossart.

**Comment by xiaoxiang781216** [comment]
Fix typo in print_time

File: drivers/rpmsg/Kconfig (hunk 13 above)
Comment (plbossart) [reply drivers/rpmsg/Kconfig:27]: Humm, that's ambitious. the platforms I know have a lot of restrictions on how clock changes are done, and to synchronize the operations between what the RTOS requests and the kernel does is no small task.

File: drivers/rpmsg/Kconfig (hunk 14 above)
Comment (plbossart) [reply drivers/rpmsg/Kconfig:67]: I understand this example.
I get how this rpmsg mechanism might be used to request services from the kernel, e.g. the clocks - with the caveat that this is far from trivial and might have security impacts.

However I don't understand how this might be used for audio stuff typically handled by the SOF firmware (pipelines, processing, etc). On most platforms I am aware of, the host runs Linux and there is a small DSP/MCU taking care of audio tasks. That latter unit is under complete control of the host and will run some sort of RTOS/scheduler along with the SOF firmware.

In short, I don't see what SOF and this rpmsg IPC have to do with each other.
Not trying to be painful, just completely missing the concept.

File: drivers/rpmsg/rpmsg_syslog.c (hunk 10 above)
Comment (plbossart) [reply drivers/rpmsg/rpmsg_syslog.c]: just add a comment to the NOTIMESTAMP definition that it's to make a checkpatch warning go away.
I just don't why this argument is necessary though, still missing the idea.

File: drivers/rpmsg/rpmsg_core.c (hunk 12 above)
Comment (plbossart) [reply drivers/rpmsg/rpmsg_core.c:88]: if you did reimplement you should add a comment to avoid confusing the reader.

I will not provide any legal advice or statement, just highlighting the fact that when reusing concepts you need to look into license opens, e.g. run this code through accepted tools to avoid GPL violation issues.

File: drivers/rpmsg/rpmsg_syslog.c (hunk 11 above)
Comment (plbossart) [reply drivers/rpmsg/rpmsg_syslog.c]: ok, adding a comment would be good to explain what you are doing here.

File: drivers/rpmsg/rpmsg_core.c (hunk 12 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/rpmsg_core.c:88]: Ok, I will update the commit message to reflect the truth.

File: drivers/rpmsg/Kconfig (hunk 14 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/Kconfig:67]: Yes, rpmsg-usrsock isn't directly relate to audio stuff, but other driver could be used in audio:
1.rpmsg-syslog could replace sof\src\include\sof\trace.h, the benifit is that:
  a.We can log the full message instead three chars
  b.We can directly see the log as normal without rmbox
  c.The implementation is more easy to port to other platform
2.With rpmsg-tty, any terminal tool from host side could connect the RTOS shell running on MCU/DSP. this is very useful for developer to debug the problem.
3.rpmsg-hostfs make MCU/DSP access the host file system, for audio:
  a.Load the tuning parameter from file.
  b.Save the tuning parameter to file during manufacturing.
  c.Dump audio pcm to file for debugging and tuning.
  d.Load audio file from host for testing
  e.Dynamiclly load module(.so) from host.
4.rpmsg-clk could control the clock on the host side.
And this is just the first patchset, in the next two patchset, we will show:
1.How to load the firmware from Linux kernel, this part normally platform specific, but which is totally decouple from rpmsg driver layer.
2.Our topology implemenation base on rpmsg framework which show that we can just focus on audio part, and decouple from the firmware loader and common service.
Since SOF is target for all vendor, it's important to ensure that audio stuff as platform independent as possible from both the kernel side and RTOS side.
With remoteproc/rpmsg framework, each vendor just need implement the remoteproc interface in drivers/remoteproc, all other part could be reused without any modification.
The flexible framework also make it's very easy to support the combined use case which is also important since one MCU/DSP may take several role(e.g. audio plus sensor hub).

**Comment by lgirdwood** [comment]
@xiaoxiang781216 have you had any feedback from the rpmsg maintainer yet ? I would definitely send to the list below
```
REMOTE PROCESSOR MESSAGING (RPMSG) SUBSYSTEM
M:	Ohad Ben-Cohen <ohad@wizery.com>
M:	Bjorn Andersson <bjorn.andersson@linaro.org>
L:	linux-remoteproc@vger.kernel.org
T:	git git://git.kernel.org/pub/scm/linux/kernel/git/ohad/rpmsg.git
S:	Maintained
F:	drivers/rpmsg/
F:	Documentation/rpmsg.txt
F:	include/linux/rpmsg.h
F:	include/linux/rpmsg/
```

File: drivers/rpmsg/Kconfig (hunk 13 above)
Comment (xiaoxiang781216) [reply drivers/rpmsg/Kconfig:27]: Yes, it's not a trival task. Our clk driver expert spend several months to achieve the  above picture.

**Comment by xiaoxiang781216** [comment]
@lgirdwood, I didn't talk to rpmsg mantainer before, we can send the patch to Bjorn after you finish the review?
> @xiaoxiang781216 have you had any feedback from the rpmsg maintainer yet ? I would definitely send to the list below
> 
> ```
> REMOTE PROCESSOR MESSAGING (RPMSG) SUBSYSTEM
> M:	Ohad Ben-Cohen <ohad@wizery.com>
> M:	Bjorn Andersson <bjorn.andersson@linaro.org>
> L:	linux-remoteproc@vger.kernel.org
> T:	git git://git.kernel.org/pub/scm/linux/kernel/git/ohad/rpmsg.git
> S:	Maintained
> F:	drivers/rpmsg/
> F:	Documentation/rpmsg.txt
> F:	include/linux/rpmsg.h
> F:	include/linux/rpmsg/
> ```

**Comment by plbossart** [comment]
My understanding of the process is 
a) the patches are submitted through the RPMSG subsystems
b) the patches are backported to topic/sof-dev
c) Additional patches making use of this framework with SOF are submitted.
For now I am not sure if there is any ask and further reviews needed here?

**Comment by xiaoxiang781216** [comment]
I will try to provide the next two patchset by this or next week.
> My understanding of the process is
> a) the patches are submitted through the RPMSG subsystems
> b) the patches are backported to topic/sof-dev
> c) Additional patches making use of this framework with SOF are submitted.
> For now I am not sure if there is any ask and further reviews needed here?

**Comment by sofci** [comment]
Can one of the admins verify this patch?

**Comment by lgirdwood** [comment]
@sofci this patch is being sent to the upstream mailing list by @xiaoxiang781216

**Comment by xiulipan** [comment]
@lgirdwood 
sofci message is some access control for the PR. Will remove that.

**Comment by plbossart** [comment]
I haven't seen any activity so closing for now. I'd like to count in-flight PRs with my two hands.
